feat: Third Person Character Controller integration with custom IP UI

fernando-cortez opened the pull request:
This PR adds a new Player Prefab for NetworkManager, based off of the Third Person Controller Starter Assets. Previous functionality of movement, collisions, and picking up is intact. The new Input System has been integration for input polling. ClientNetworkAnimator enables the animations of the player to be client-driven.

Included as well is an overhaul on the UI to now include custom IPs, using UI Toolkit.

[MTT-5013](https://jira.unity3d.com/browse/MTT-5013)

Package Manager manifest will target the new Boss Room release containing the new Utilities release, once that is out!

Testing instructions:
* Create a project clone via ParrelSync, or create a build of the game. To test the game via the editor, open up the Bootstrap scene and hit Play.
* Have one instance of the game host, and have other instances join the host.
* Here, validate that all clients see animations being played out from every other client. Test picking up ingredients with E.

LPLafontaineB commented on Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:
```diff
@@ -36,60 +41,50 @@ public override void OnNetworkSpawn()
         if (!IsOwner)
         {
-            m_Camera.gameObject.SetActive(false);
+            m_ThirdPersonController.enabled = false;
```
Do we need to set it false here too since it is already set to false on awake?

LPLafontaineB commented on Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:
```diff
@@ -36,60 +41,50 @@ public override void OnNetworkSpawn()
+        }
+        else
+        {
+            // todo use non-alloc, don't do the below at home
```
should we do this todo now while we are here?

LPLafontaineB commented on Basic/ClientDriven/Assets/Scripts/HostJoinUI.cs:
```diff
@@ -0,0 +1,115 @@
+    public void StartHost(EventBase obj)
+    {
+        SetUtpConnectionData();
+        var result= NetworkManager.Singleton.StartHost();
```
```suggestion
        var result = NetworkManager.Singleton.StartHost();
```
Same thing in the other methods below

LPLafontaineB reviewed (commented):
Mostly minor nitpicks in code. Otherwise I see an issue where when we drop an object it disappears into the ground. Also another really nitpicky comment, but since we can now jump, maybe the "counter" should be a bit higher so that we don't notice the invisible wall too much? It might be something for a later PR though

RikuTheFuffs commented:
@fernando-cortez would it be possible to have some testing instructions listed in the PR? I'd like to test it locally, but dont't really know if I should do something in particular to set up the environment or if there's a specific key combiantion to test

fernando-cortez commented:
> @fernando-cortez would it be possible to have some testing instructions listed in the PR? I'd like to test it locally, but dont't really know if I should do something in particular to set up the environment or if there's a specific key combination to test

Absolutely, I'll include some test instructions in the summary!

RikuTheFuffs commented on Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:
```diff
@@ -9,17 +9,12 @@
-        m_Client = GetComponent<ClientPlayerMove>();
-    }
+    
+    NetworkObject m_PickedUpObj;
```
as per our [coding conventions](https://github.cds.internal.unity3d.com/unity/unity-meta/blob/stable/ReferenceSource/CSharp/Assets/CSharpReference.cs), we should not use abbreviations in code, unless we're dealing with acronyms

RikuTheFuffs commented on Basic/ClientDriven/Assets/Textures/backgroundGlow.png.meta:
```diff
@@ -0,0 +1,146 @@
+  serializedVersion: 11
+  mipmaps:
+    mipMapMode: 0
+    enableMipMap: 1
```
As this is used in a screen-overlay UI mipmaps should be disabled (saves an amount of memory equal to ~33% of the texture's weight)

RikuTheFuffs reviewed (changes requested):
1. There seem to be some warnings related to the Chintzy font (no stack trace). Do you know what's causing them? Did you try addressing them already? You can let them pop up by dropping a colored object in one of the circles of the same color as the object

![image](https://user-images.githubusercontent.com/27730963/202147712-84b0c105-616e-4c36-aff0-99aa4bcdf56e.png)

2. There's something wrong with one of the network variables:

![image](https://user-images.githubusercontent.com/27730963/202147510-ce97c826-7b83-42e2-bda2-c861a8805108.png)

3. Going out of play mode after playing for a while thows:


```
Type of instance in array does not match expected type
UnityEngine.Object:Destroy (UnityEngine.Object)
Unity.Netcode.NetworkSpawnManager:OnDespawnObject (Unity.Netcode.NetworkObject,bool) (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Spawning/NetworkSpawnManager.cs:873)
Unity.Netcode.NetworkSpawnManager:DespawnAndDestroyNetworkObjects () (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Spawning/NetworkSpawnManager.cs:699)
Unity.Netcode.NetworkManager:ShutdownInternal () (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Core/NetworkManager.cs:1439)
Unity.Netcode.NetworkManager:OnDestroy () (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Core/NetworkManager.cs:1331)
Unity.Netcode.NetworkManager:OnApplicationQuit () (at Library/PackageCache/com.unity.netcode.gameobjects@1.1.0/Runtime/Core/NetworkManager.cs:1325)

```

RikuTheFuffs commented:
Also, I'm not sure if this is on purpose, but there seem to be some interactivity problems:

1. Dropping an object outside of a circle by pressingthe E button again makes the object fall under the ground
2. Under certain circumstances, I was able to grab objects very far away and to make them follow my character's rotation. I didn't manage to repro this consistently, but spamming the E button probably helps
3. Sometimes, I was able to pick up objects even when I had nothing around me. Pressing E would just make something pop up on me, as if I summoned it out of thin air

fernando-cortez commented on Basic/ClientDriven/Assets/Scripts/HostJoinUI.cs:
```diff
@@ -0,0 +1,122 @@
+
+    void SetUtpConnectionData()
+    {
+        var sanitizedIPText = Sanitize(m_IPAddressTextField.text);
+        var sanitizedPortText = Sanitize(m_PortTextField.text);
+
+        ushort.TryParse(sanitizedPortText, out var port);
```
Would you be able to direct me to the specific bit where it specifies this? We've used vars even without context inside of Boss Room so I'm curious about this.

RikuTheFuffs replied:
Oh, you actually made me realize that the documentation says (at [this line](https://github.cds.internal.unity3d.com/unity/unity-meta/blob/c0eeb02f04bdd06f0aedc0eee78e67ce20786c5d/ReferenceSource/CSharp/Assets/CSharpReference.cs#L223)) that you should use var when the type is explicit, but says nothing about what to do when it is not. My bad! I guess what's written here is fine, then.

fernando-cortez commented:
@RikuTheFuffs addressed the comments you've raised and left feedback on others as well. The interactivity of the sample will feel a lot better, with non-allocating boxcasts from the players direction now determining whether an object should be picked up. Picked up objects also now don't fall through the floor 😅

RikuTheFuffs commented on Basic/ClientDriven/Assets/Scripts/ServerIngredient.cs:
```diff
@@ -7,14 +8,19 @@ public class ServerIngredient : ServerObjectWithIngredientType
 
 public class ServerObjectWithIngredientType : NetworkBehaviour
 {
+    [SerializeField]
+    public NetworkVariable<IngredientType> CurrentIngredientType;
```
public fields are automatically serialized, so the SerializeField here is redundant (or even dangerous if it causes the serialization o happen twice). I'd suggest to either use `internal` + `SerializeField`, or to drop the `SerializeField`

SamuelBellomo replied:
I'd be surprised if public + SerializeField cause the serialization to happen twice, any link to that?

RikuTheFuffs replied:
I remember somebody from the engine team pointing that out long time ago to me in a PR, but tbh I don't remember for which Unity version that observation was valid (and if it was by design or a bug of a specific major version)

RikuTheFuffs replied:
just re-asked [here](https://unity.slack.com/archives/C9SQHJGN6/p1669907293851319) to be sure

RikuTheFuffs replied:
got an answer: "there's no impact in using both [SerializeField] + public". Sorry for the false alarm!

RikuTheFuffs commented on Basic/ClientDriven/Assets/Scripts/ServerScoreReplicator.cs:
```diff
@@ -1,9 +1,10 @@
-    private NetworkVariable<int> m_ReplicatedScore = new NetworkVariable<int>();
+    NetworkVariable<int> m_ReplicatedScore = new NetworkVariable<int>();
+
+    public NetworkVariable<int> replicatedScore => m_ReplicatedScore; 
```
Nitpick: being a property, it should be `ReplicatedScore` (link [here](https://github.cds.internal.unity3d.com/unity/unity-meta/blob/c0eeb02f04bdd06f0aedc0eee78e67ce20786c5d/ReferenceSource/CSharp/Assets/CSharpReference.cs#L194))

RikuTheFuffs reviewed (approved):
I just retested it and it looks great! Apart from the `[SerializeField]` -related note i left, everything looks good from my side

SamuelBellomo commented on Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:
```diff
@@ -1,6 +1,9 @@
+using System;
```
Adding a note here, not linked to this file.
wanna add an entry to the changelog please?

SamuelBellomo commented on Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:
```diff
@@ -36,60 +44,55 @@ public override void OnNetworkSpawn()
+        m_PlayerInput.enabled = true;
+        m_ThirdPersonController.enabled = true;
 
-    // DOC START HERE
```
Was the doc updated?

fernando-cortez replied:
Right. If it makes sense to you, I've created a task to update ClientDriven doc & create a PR with the changes here: [MTT-5106](https://jira.unity3d.com/browse/MTT-5106)

SamuelBellomo replied:
When is that fix gonna land? Cause as soon as we release this PR, docs will break.

fernando-cortez replied:
Created this [docs PR](https://jira.unity3d.com/browse/MTT-5013) with scripting changes!

SamuelBellomo replied:
you mean this? https://github.com/Unity-Technologies/com.unity.multiplayer.docs/pull/865/files

SamuelBellomo commented on Basic/ClientDriven/Assets/Scripts/ServerIngredientSpawner.cs:
```diff
@@ -27,19 +33,24 @@ public override void OnNetworkSpawn()
                 ingredient.NetworkObject.Spawn();
+                ingredient.CurrentIngredientType.Value = (IngredientType)m_RandomGenerator.Next((int)IngredientType.MAX);
             }
+            m_SpawnWaves++;
```
I don't see spawnWaves being reset anywhere or decremented. Does it mean if players place all ingredients in the circles, there won't be new ingredients?
(That can work, I don't mind, just checking if this is the intended behaviour)

SamuelBellomo commented on Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:
```diff
@@ -9,17 +9,12 @@
+    
+    NetworkObject m_PickedUpObj;
 
     // DOC START HERE
```
inherit doc is for xml doc generation. The above is a note for developers, telling us to go to our docs website and update the line numbers that reference the samples. There's (unfortunately) no automation with the above "DOC START HERE" :(

SamuelBellomo commented on Basic/ClientDriven/Assets/Scripts/ServerPlayerMove.cs:
```diff
@@ -31,42 +35,66 @@ public override void OnNetworkSpawn()
+    [ServerRpc]
+    public void DropObjServerRpc()
+    {
+        m_DropRequested = true;
```
When we were looking at this, I thought this didn't fix the issue, was that it in the end? Or did you forget to reverse that change?

SamuelBellomo replied:
I guess because RPCs can happen in update, deferring physics changes to FixedUpdate makes sense still... nvm the above comment.
I'm still curious though whether that fixed the issue or not

SamuelBellomo replied:
If it did, it'd be great to comment this and produce some edu comments.

fernando-cortez replied:
When we were trying things out, the changes we made were client side on sending RPCs to the server. This modification on the server side of code resolved the issue. 
Though what's curious to me is moving this code to a coroutine and waiting for the next FixedUpdate did not produce the same results. I want to follow up with SDK on this offline, but this will absolutely go into the updated doc, as a new section.

SamuelBellomo replied:
before telling our users they should do this, it'd be great if we understood exactly why (and make sure this isn't an issue with the SDK that we're just building a workaround for --> which would need to be reverted a month or two later...)

fernando-cortez replied:
This has been since removed. The previous functionality of dropping an ingredient overhead was restored. This issue will be addressed (if need be) in the follow-up PR that integrates the pickup animations.

SamuelBellomo commented on Basic/ClientDriven/Assets/StarterAssets/ThirdPersonController/Scripts/ThirdPersonController.cs:
```diff
@@ -96,7 +96,6 @@ public class ThirdPersonController : MonoBehaviour
         private int _animIDGrounded;
         private int _animIDJump;
         private int _animIDFreeFall;
-        private int _animIDMotionSpeed;
```
little trick for future proofing: make sure you gate all your changes to Starter assets with something like // MTT CHANGE START and // MTT CHANGE STOP something like that. This way when merging new StarterAsset content later, you'll know which parts you did change yourself and which are just merge changes.
Hopefully once we'll update Starter assets with NGO we'll have less of those issues.

SamuelBellomo commented on Basic/ClientDriven/Assets/UI Toolkit/2DSpaceShooterUIStyleSheet.uss:
```diff
@@ -0,0 +1,241 @@
+* {
```
I see 2D space shooter in the title, I'm assuming this was copied from the other bitesize?
I'm wondering if couldn't just create ourselves a common package between bitesize projects for things like these. Were there many changes to that sheet vs the space shooter?
might be something to sync with Jil

SamuelBellomo commented on Basic/ClientDriven/Packages/manifest.json:
```diff
@@ -6,7 +6,7 @@
     "com.unity.ide.vscode": "1.2.5",
     "com.unity.inputsystem": "1.4.3",
-    "com.unity.multiplayer.samples.coop": "https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop.git?path=/Packages/com.unity.multiplayer.samples.coop",
+    "com.unity.multiplayer.samples.coop": "https://github.com/Unity-Technologies/com.unity.multiplayer.samples.coop.git?path=/Packages/com.unity.multiplayer.samples.coop#a63fde7fbdfdbd601bae9cc42d93b7e6bc9d49c8",
```
adding a comment here to block this until it's merged boss room side.

fernando-cortez replied:
I'll leave your comment open until the Boss Room update has been merged.

SamuelBellomo commented on CHANGELOG.md:
```diff
@@ -1,5 +1,15 @@
+
+#### Changed
+- Updated to Netcode for GameObjects 1.1.0
+- The sample has been refactored to feature a PlayerPrefab inside NetworkManager that is a networked variant of a prefab from [Unity's Starter Assets](https://assetstore.unity.com/packages/essentials/starter-assets-third-person-character-controller-196526). Other improvements include:
```
It'd be great if we could discuss some of the bug fixes and improvements to Starter Assets (or link to the doc that does) that users would need to do on their side to make this work.

fernando-cortez replied:
I've created a Jira task [here](https://jira.unity3d.com/browse/MTT-5179) to track this doc.

fernando-cortez replied:
Closed the task, per offline discussion.

SamuelBellomo commented on Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:
```diff
@@ -45,6 +49,8 @@ public override void OnNetworkSpawn()
         if (!IsOwner)
         {
             enabled = false;
+            m_CharacterController.enabled = false;
```
mind adding a quick explanation on the why we did this? Those lines are part of the "magic" that makes Character Controller work, this should be highlighted.

SamuelBellomo commented on Basic/ClientDriven/Assets/Prefabs/Ingredient.prefab:
```diff
@@ -641,9 +655,9 @@ MonoBehaviour:
   SyncRotAngleX: 1
```
I can't remember, was looking up and down replicated as well? Or just character position and animations?

fernando-cortez replied:
Synced offline. Will look into adding looking rotation to a follow-up update.

SamuelBellomo commented on Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:
```diff
@@ -37,6 +37,12 @@ void Awake()
         Cursor.lockState = CursorLockMode.Locked;
         Cursor.visible = false;
         
+        // ThirdPersonController & CharacterController are enabled only on owning clients. Ghost clients have these two 
```
```suggestion
        // ThirdPersonController & CharacterController are enabled only on owning clients. Ghost player objects have these two 
```

SamuelBellomo commented on Basic/ClientDriven/Assets/Scripts/ClientPlayerMove.cs:
```diff
@@ -37,6 +37,12 @@ void Awake()
         
+        // ThirdPersonController & CharacterController are enabled only on owning clients. Ghost clients have these two 
+        // components disabled, and will enable a CapsuleCollider. Per the CharacterController documentation: 
+        // https://docs.unity3d.com/Manual/CharacterControllers.html, a Character controller can push rigidbody
```
oh nice! good to know
@RikuTheFuffs @chrispope hopefully this will address most (?) issues raised by users on character controller usage? I'm not aware of other issues, if there's any other known issues, please let us know

RikuTheFuffs replied:
Other issues I read about were related to the `PlayerInput` component being enabled by default, which caused its `Actions` field to be initialized with some weird new instance of `Input Action Asset`.
An easy solution to that is to disable the component by default, and to enable it only on the localPlayer

![image](https://user-images.githubusercontent.com/27730963/205088593-5fc66fb9-bdc2-42a7-aefa-5d2ffd12370c.png)

fernando-cortez replied:
That's a comment on proper usage, @RikuTheFuffs? Just checking since that component is disabled on the PlayerArmature_Networked player prefab, and only enabled on spawn for the owning player.

SamuelBellomo replied:
Might worth to still note this in our guide, to make sure inputs are handled only by owner clients and not other clients

fernando-cortez replied:
Added [MTT-5190](https://jira.unity3d.com/browse/MTT-5190) as a follow up to the ClientDriven doc.

SamuelBellomo commented on Basic/ClientDriven/Assets/StarterAssets/ThirdPersonController/Scripts/ThirdPersonController.cs:
```diff
@@ -96,7 +96,13 @@ public class ThirdPersonController : MonoBehaviour
+        // Reason: the Third Person Controller Animator uses a speed multiplier on the player's walk/run blend tree 
+        // state, to apply speed based on a player's input. On ghost clients, this component is disabled on ghost
+        // clients, meaning this parameter is at it's default value, and the walk/run blend tree will appear stuck.  
+        // To circumvent that, the walk/run blend tree state is not multiplied by any game-side code.
```
mmm so wait, doesn't that reduce the amount of functionality from the original Starter Assets? If users want to change the character's speed, with this change here the animations wouldn't react to it? Why aren't we trying to sync that magnitude info? Or deduce it from the character's current speed?

fernando-cortez replied:
Followed up offline and this has since been re-enabled!
